advection_test/functions.py

```diff
@@ -0,0 +1,1302 @@
+        tracers,
+    )
+
+    return tracAdv
+
+
+def prepare_everythingForAdvection(
+    configuration,
+    uc,
+    vc,
+    delp,
+    tracers,
+    density,
+    dimensions,
+    units,
+    origins,
+    backend,
+    fvt_dict,
+    dt_acoustic=300,
+):
+    """
+    Use: fluxPrep, tracAdv = prepare_everythingForAdvection(
+        configuration, uc, vc, delp, tracers, density, dimensions, units,
+        origins, backend, fvt_dict, dt_acoustic=300)
+
+    Inputs:
+    - configuration: Dict{'stencil_factory', 'grid_data', 'damping_coefficients'}
+    - uc, vc: x- and y- winds on C-grid
+    - delp: pressure thickness of layer
+    - tracers: Dict{'smoke}
+    - density: (kg/m3 for air)
+    - dimensions: Dict{'nx', 'ny', 'nz', 'nx1', 'ny1', 'nxhalo', 'nyhalo'}
+    - units: Dict{'areaflux', 'courant', 'mass', 'wind'}
+    - origins: Dict{'compute_3d'}
+    - backend: 'numpy'
+    - fvt_dict: Dict{'grid_type', 'hord'}
+    - dt_acoustic: acoustic time step (default = 300 seconds)
+
+    Outputs:
+    - fluxPrep: Dict{'crx', 'cry', 'mfxd', 'mfyd', 'ucv', 'vcv', 'xaf', 'yaf'}
+        - crx, cry are Courant numbers
+        - ucv, vcv are the contravariant winds
+        - xaf, yaf are fluxes of area (dx*dy but incorporating wind speed)
+        - mfxd, mfyd are mass (volume) fluxes incorporating wind speed
+    - tracAdv: class instance that performs tracer advection
+    """
+
+    fluxPrep = run_finiteVolumeFluxPrep(
+        configuration,
+        uc,
+        vc,
+        delp,
+        density,
+        dimensions,
+        units,
+        origins,
+        backend,
+        dt_acoustic,
+    )
+
+    tracAdv = build_tracerAdvection(configuration, fvt_dict, tracers)
+
+    tracAdv_data = {
+        "tracers": tracers,
+        "delp": delp,
+        "mfxd": fluxPrep["mfxd"],
+        "mfyd": fluxPrep["mfyd"],
+        "crx": fluxPrep["crx"],
+        "cry": fluxPrep["cry"],
+    }
+
+    return tracAdv_data, tracAdv
+
+
+def run_advectionStepWithReset(
+    tracAdv_dataInit, tracAdv_data, tracAdv, dt, mpi_rank=None
+):
+    """
+    Use: tracAdv_data = run_advectionStepWithReset()
+
+    Inputs:
+    - tracAdv_dataInit: Dict{'tracers', 'delp', 'mfxd', 'mfyd', 'crx', 'cry'}
+                        of values to be reset to after each advection step
+    - tracAdv_data: Dict{'tracers', 'delp', 'mfxd', 'mfyd', 'crx', 'cry'}
+    - tracAdv: class instance of TracerAdvection
+    - dt: time step in seconds
+    - mpi_rank: if 0, prints differences from timestep, initial condition
+
+    Outputs:
+    - tracAdv_data: updated fields (of only tracer)
+    """
+
+    tmp = cp.deepcopy(tracAdv_data["tracers"])  # pre-advection tracer state
+
+    tracAdv(
+        tracAdv_data["tracers"],
+        tracAdv_data["delp"],
+        tracAdv_data["mfxd"],
+        tracAdv_data["mfyd"],
+        tracAdv_data["crx"],
+        tracAdv_data["cry"],
+        dt,
+    )
+
+    tracAdv_data["delp"] = cp.deepcopy(tracAdv_dataInit["delp"])
```

Comment:
explain why we are doing this

Reply:
Just in case - since all the others need to be force-reset.